### PR TITLE
MRG, FIX, ENH: overhaul to_data_frame method

### LINKED ISSUE #7179: DOC: epochs.to_data_frame doc error
#### Describe the bug
Docs say that the `index` parameter should be `None` or a tuple of strings. Passing tuples yields a `KeyError`, but passing a list works as intended.

#### Steps to reproduce
with any epochs object:
```
epochs.to_data_frame(index=('time', 'condition'))
```

#### Expected results
A pandas dataframe with MultiIndex levels "time" and "condition" and a column for epoch number.

#### Actual results
<details>
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/opt/miniconda3/envs/mnedev/lib/python3.7/site-packages/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2656             try:
-> 2657                 return self._engine.get_loc(key)
   2658             except KeyError:

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()

KeyError: ('time', 'condition')

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
<ipython-input-11-b7b430af7042> in <module>
----> 1 df = epochs.to_data_frame(index=('time','condition'))

/opt/mne/python/mne/io/base.py in to_data_frame(self, picks, index, scaling_time, scalings, copy, start, stop, long_format)
    206             if 'time' in index and not long_format:
    207                 _set_pandas_dtype(df, ['time'], np.int64)
--> 208             df.set_index(index, inplace=True)
    209         if all(i in default_index for i in index):
    210             if isinstance(self, _BaseSourceEstimate):

/opt/miniconda3/envs/mnedev/lib/python3.7/site-packages/pandas/core/frame.py in set_index(self, keys, drop, append, inplace, verify_integrity)
   4176                 names.append(None)
   4177             else:
-> 4178                 level = frame[col]._values
   4179                 names.append(col)
   4180                 if drop:

/opt/miniconda3/envs/mnedev/lib/python3.7/site-packages/pandas/core/frame.py in __getitem__(self, key)
   2925             if self.columns.nlevels > 1:
   2926                 return self._getitem_multilevel(key)
-> 2927             indexer = self.columns.get_loc(key)
   2928             if is_integer(indexer):
   2929                 indexer = [indexer]

/opt/miniconda3/envs/mnedev/lib/python3.7/site-packages/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2657                 return self._engine.get_loc(key)
   2658             except KeyError:
-> 2659                 return self._engine.get_loc(self._maybe_cast_indexer(key))
   2660         indexer = self.get_indexer([key], method=method, tolerance=tolerance)
   2661         if indexer.ndim > 1 or indexer.size > 1:

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()

KeyError: ('time', 'condition')
</details>

#### Additional information

<details>
Platform:      Linux-4.15.0-72-generic-x86_64-with-debian-buster-sid
Python:        3.7.3 | packaged by conda-forge | (default, Mar 27 2019, 23:01:00)  [GCC 7.3.0]
Executable:    /opt/miniconda3/envs/mnedev/bin/python
CPU:           x86_64: 8 cores
Memory:        7.6 GB

mne:           0.20.dev0
numpy:         1.16.4 {blas=blas, lapack=blas}
scipy:         1.3.0
matplotlib:    3.1.0 {backend=Qt5Agg}

sklearn:       0.21.2
numba:         Not found
nibabel:       2.4.1
cupy:          Not found
pandas:        0.24.2
dipy:          0.16.0
mayavi:        4.7.1.dev0 {qt_api=pyqt5, PyQt5=5.9.2}
pyvista:       0.20.4
vtk:           8.1.2
</details>

### PR DESCRIPTION
**Pull request by drammock**
closes #7179 

- [x] minor refactor of `_check_pandas_index_arguments`
- [x] fix `to_data_frame()` docstring re: tuples vs lists
- [x] make `index=None` work as described
- [x] fix logic when setting title of columns
- [x] provide different options for time conversion: keep as float, or convert to integer milliseconds, pd.Timedelta, or (raw only) pd.Timestamp
- [x] make the datetime/timedelta change work with `long_format=True`
- [x] improve tests

cc @jona-sassenhagen @dengemann @choldgraf @larsoner @agramfort

### CODE REVIEW

**Comment by larsoner** [comment]
> the changes that convert time indexes to pandas.DatetimeIndex or pandas.TimedeltaIndex (instead of an int64 index in milliseconds) are API changes that I think are sensible but may not be agreeable to everyone

Can we provide `index='ms'` as a backward-compatible option and by deprecation change the default from `ms` to `time`?

**Comment by drammock** [comment]
> > the changes that convert time indexes to pandas.DatetimeIndex or pandas.TimedeltaIndex (instead of an int64 index in milliseconds) are API changes that I think are sensible but may not be agreeable to everyone
> 
> Can we provide `index='ms'` as a backward-compatible option and by deprecation change the default from `ms` to `time`?

I could add `'ms'` as a supported string argument to `index`.  But:

- the current default is `index=None`, which (after the bugfix in b801b7c) won't put time into the index anyway, so I'm not sure how we would do a deprecation cycle.
- when time is not in the index, there is no need to convert to milliseconds and cast to integer. Yet for some reason the old behavior was to convert to milliseconds and round (without casting to integer). This is really weird behavior in my opinion, and I consider it a second bugfix (in c53fa0e) to simply leave times alone (as floats, in seconds) if they're not being put into the index.

**Comment by larsoner** [comment]
Okay let's just document it well and consider it a bugfix. My guess is that these bits of code are not extensively used

**Comment by drammock** [comment]
> My guess is that these bits of code are not extensively used

That's my guess too.

**Comment by drammock** [comment]
@larsoner  I think this is ready for review.  TLDR:

- the PR is backward compatible in terms of *default behavior*, except for whether columns end up in the index or not (which was previously a bug IMO; see below).
- The PR is not backward compatible in terms of *method parameters* (`scale_time` is gone, `time_format` is added).
- `to_data_frame` is no longer a `Mixin` with complicated logic to triage the different instance types; it has separate definitions for `Raw`, `Epochs`, `Evoked`, and `SourceEstimate` classes, and some share `utils` functions to make it DRY.
- all of the instance methods now give better control over how the time variable is transformed (if at all)

More details:

- `<instance>.to_data_frame(index=None)` now works as advertised: the DataFrame will have a sequential integer index (the Pandas default) and the indicator variables will all be in their own columns.
- There's a new option `time_format` that can be `None` (keep time as float), `'ms'` (convert to integer milliseconds), `timedelta` (convert to `pd.Timedelta`), and `datetime` (convert to `pd.Timestamp`; only works for Raw objs, and accounts for `meas_date` and `raw.first_time`).  This last option is probably not widely useful, but was not hard to add and is in there for completeness.
- The old `scaling_time=1e3` API element is gone. Default behavior is unchanged, and if users ever passed anything other than `1e3`, they can still get that result by passing `time_format=None` and then scaling the resulting dataframe column after the fact.  Currently I haven't done any deprecation code for the `scaling_time` parameter; LMK if you think a deprecation cycle is needed there.
- The above approach separates the question of time format conversion from whether time is in the DataFrame index or not.  It also makes it possible to do a deprecation cycle on the default value of `time_format`... I still think `None` is a more sensible default than `'ms'`, but we can do the deprecation in a separate PR easily enough, and I'm personally content just to have the option of passing `time_format=None`.

**Comment by larsoner** [comment]
Your described current approach sounds good to me, will look in depth tomorrow

**Comment by drammock** [comment]
The Azure failures are puzzling me; these all pass locally, and Travis is happy:

```
pytest mne/tests/test_epochs.py -k test_to_data_frame
pytest mne/tests/test_evoked.py -k test_to_data_frame
pytest mne/tests/test_source_estimate.py -k test_to_data_frame
pytest mne/io/fiff/tests/test_raw_fiff.py -k test_to_data_frame
pytest mne/io/edf/tests/test_edf.py -k test_to_data_frame
```

Looks like it's an error with `isinstance` testing against `np.int64`.  Locally it was failing when I tested against plain `int` which is why I switched the test to int64...  is there a better way to do this kind of type checking?

**Comment by dengemann** [comment]
Sounds good, l’ll take a look asap and check out how this interacts with R.

**Comment by agramfort** [comment]
this issue of int coding has been appearing in the past. See for example in
here:


https://github.com/mne-tools/mne-python/issues?utf8=%E2%9C%93&q=is%3Aissue+np.int64

>

File: mne/utils/dataframe.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            data[:, idx] *= scaling
+    return data
+
+
+def _convert_times(inst, times, time_format):
+    """Convert vector of time in seconds to ms, datetime, or timedelta."""
+    # private function; pandas already checked in calling function
+    from pandas import to_timedelta
+    if time_format == 'ms':
+        times = np.round(times * 1e3).astype(int)
```
Comment (larsoner) [inline mne/utils/dataframe.py]: This might need to be `np.int64` for example

**Comment by drammock** [comment]
Rebased to fix a doc building issue that was fixed elsewhere.  @dengemann do you still have time to review this?  There's a tutorial PR #7180 that is waiting for it.

@jona-sassenhagen @choldgraf any interest in weighing in?

**Comment by dengemann** [comment]
Tests pass on MNE-R and examples build.  But as we just discovered, the time_format  options can cause trouble and hit untested territory.
Let’s do an issue in mne-r and in case of doubt break things to move on here. We’ll then patch it up over there.

Denis

**Comment by drammock** [comment]
Summarizing an offline conversation with @dengemann: this PR doesn't *break* MNE-R, but the rewritten to_data_frame methods don't work cleanly, and some options are causing segfaults.  We'll investigate for a day or two to see how hard it will be to fix it.

**Comment by dengemann** [comment]
Too fast, I had  a path issue and did not actually run your code …
On MNE master MNE-R  is fine, on this branch it  produces 1failing test.
Can you reproduce?

**Comment by dengemann** [comment]
@drammock it looks very clean to me. Thank you for this very nice work and the testing against mne-r ! I'd say let's move on.

**Comment by larsoner** [comment]
@drammock this appears to have broken three examples:

https://circleci.com/gh/mne-tools/mne-python/17989

Can you look into it?

**Comment by drammock** [comment]
Will do.

**Pull request merged**